core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -140,6 +144,16 @@ private[spark] class CoarseGrainedExecutorBackend(
       if (executor == null) {
         exitExecutor(1, "Received LaunchTask command but executor was null")
       } else {
+        if (decommissioned) {
+          logError("Asked to launch a task while decommissioned.")
+          driver match {
+            case Some(endpoint) =>
```

Comment:
I think that instead of doing this here, it should be done in `onStart` where the driver reference is created. That means the decommission message is sent to the driver as soon as possible after the signal arrives, instead of waiting for the driver to try to use the executor for something.

(That also means this block can go away and you can just keep the log message in `Executor.scala`.)

Reply:
So what about when we are scaling down after the driver reference is created?

Reply:
https://docs.google.com/document/d/1xVO1b6KAwdUhjEJBolVPl9C6sLj7oOveErwDSYdT-pE/edit?disco=AAAAI73a0FM
I have marked a comment on this in design doc as well. I think it can be handled by the driver by not allocating tasks to the executor at the first place.
When driver is aware of the possible decommission of the node, it can stop allocating tasks to this executor. A small code change in the driver's org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend.DriverEndpoint#makeOffers 
```
       // Filter out executors on decommissioning worker
        val activeExecutors = executorDataMap.filterKeys(isExecutorActive)
                                 .filter(x => !isNodeDecommissioning(x._2.executorHost))
```
For this there can be DecommissionTracker ( in the same lines as BlacklistTracker). The DT is filled when the driver is informed of the decommissioning is informed on the host. As comment on the design I have tried to elaborate the flow of populating DT.
```
  private def isNodeDecommissioning(hostname: String): Boolean = {
    decommissionTracker match {
      case None => return false
      case Some(decommissionTracker) => return decommissionTracker.isNodeDecommissioning(hostname)
    }
  }
}
```

Reply:
So @itskals on Kubernetes and in standalone the driver won't know first (and those the only parts implemented in this PR). Certainly, when we go to implement YARN support we could try and short circuit talking to the executor -- however to enable later things like migrations I'd still want to send the message.

This could make a difference when we implement YARN support and I've added a note in the design doc so that we don't skip it then.